client/batch_controller.go

```diff
@@ -0,0 +1,236 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pd
+
+import (
+	"context"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// Starting from a low value is necessary because we need to make sure it will be converged to (current_batch_size - 4).
+const defaultBestBatchSize = 8
+
+// finisherFunc is used to finish a request, it accepts the index of the request in the batch, the request itself and an error.
+type finisherFunc[T any] func(int, T, error)
+
+type batchController[T any] struct {
+	maxBatchSize int
+	// bestBatchSize is a dynamic size that changed based on the current batch effect.
+	bestBatchSize int
+
+	collectedRequests     []T
+	collectedRequestCount int
+
+	// The finisher function to cancel collected requests when an internal error occurs.
+	finisher finisherFunc[T]
+	// The observer to record the best batch size.
+	bestBatchObserver prometheus.Histogram
+	// The time after getting the first request and the token, and before performing extra batching.
+	extraBatchingStartTime time.Time
+}
+
+func newBatchController[T any](maxBatchSize int, finisher finisherFunc[T], bestBatchObserver prometheus.Histogram) *batchController[T] {
+	return &batchController[T]{
+		maxBatchSize:          maxBatchSize,
+		bestBatchSize:         defaultBestBatchSize,
+		collectedRequests:     make([]T, maxBatchSize+1),
+		collectedRequestCount: 0,
+		finisher:              finisher,
+		bestBatchObserver:     bestBatchObserver,
+	}
+}
+
+// fetchPendingRequests will start a new round of the batch collecting from the channel.
+// It returns nil error if everything goes well, otherwise a non-nil error which means we should stop the service.
+// It's guaranteed that if this function failed after collecting some requests, then these requests will be cancelled
+// when the function returns, so the caller don't need to clear them manually.
+func (bc *batchController[T]) fetchPendingRequests(ctx context.Context, requestCh <-chan T, tokenCh chan struct{}, maxBatchWaitInterval time.Duration) (errRet error) {
+	var tokenAcquired bool
+	defer func() {
+		if errRet != nil {
+			// Something went wrong when collecting a batch of requests. Release the token and cancel collected requests
+			// if any.
+			if tokenAcquired {
+				tokenCh <- struct{}{}
+			}
+			bc.finishCollectedRequests(bc.finisher, errRet)
+		}
+	}()
+
+	// Wait until BOTH the first request and the token have arrived.
+	// TODO: `bc.collectedRequestCount` should never be non-empty here. Consider do assertion here.
+	bc.collectedRequestCount = 0
+	for {
+		// If the batch size reaches the maxBatchSize limit but the token haven't arrived yet, don't receive more
+		// requests, and return when token is ready.
+		if bc.collectedRequestCount >= bc.maxBatchSize && !tokenAcquired {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
```

Comment:
Will defer be executed first?

Reply:
What do you mean? After this return, the defer function will be executed with `ctx.Err()` as `errRet`.

Reply:
got